pyvortex/src/scalar.rs

```diff
@@ -4,72 +4,75 @@
 //! :meth:`.Array.scalar_at`. They represent shared-memory views into individual values of a Vortex
 //! array.
 
-use std::sync::Arc;
-
+use pyo3::exceptions::PyValueError;
 use pyo3::prelude::*;
 use pyo3::types::PyDict;
 use vortex::buffer::{Buffer, BufferString};
-use vortex::dtype::{DType, StructDType};
-use vortex::error::vortex_panic;
-use vortex::scalar::{PValue, Scalar, ScalarValue};
+use vortex::dtype::half::f16;
+use vortex::dtype::PType;
+use vortex::error::VortexExpect as _;
+use vortex::scalar::{ListScalar, Scalar, StructScalar};
 
 pub fn scalar_into_py(py: Python, x: Scalar, copy_into_python: bool) -> PyResult<PyObject> {
-    let (dtype, value) = x.into_parts();
-    scalar_value_into_py(py, value, &dtype, copy_into_python)
-}
+    if x.is_null() {
+        return Ok(py.None());
+    }
+
+    if let Some(b) = x.as_bool_opt() {
+        return Ok(b.value().into_py(py));
+    }
+
+    if let Some(p) = x.as_primitive_opt() {
```

Comment:
might be better to do this as a `match` statement so that the compiler ensures it's exhaustive